Source/WebCore/rendering/BackgroundPainter.cpp

```diff
@@ -478,15 +478,20 @@ void BackgroundPainter::paintFillLayer(const Color& color, const FillLayer& bgLa
 
     // no progressive loading of the background image
     if (!baseBgColorOnly && shouldPaintBackgroundImage) {
-        // Multiline inline boxes paint like the image was one long strip spanning lines. The backgroundImageStrip is this fictional rectangle.
-        auto imageRect = backgroundImageStrip.isEmpty() ? scrolledPaintRect : backgroundImageStrip;
-        auto paintOffset = backgroundImageStrip.isEmpty() ? rect.location() : backgroundImageStrip.location();
-        auto geometry = calculateBackgroundImageGeometry(m_renderer, m_paintInfo.paintContainer, bgLayer, paintOffset, imageRect, m_overrideOrigin);
+        auto paintOffset = m_imageRect ? m_imageRect->location() : rect.location();
+        auto geometry = calculateBackgroundImageGeometry(m_renderer, m_paintInfo.paintContainer, bgLayer, paintOffset, scrolledPaintRect, m_imageRect);
 
         auto& clientForBackgroundImage = backgroundObject ? *backgroundObject : m_renderer;
         bgImage->setContainerContextForRenderer(clientForBackgroundImage, geometry.tileSizeWithoutPixelSnapping, m_renderer.style().usedZoom());
 
-        geometry.clip(LayoutRect(pixelSnappedRect));
+        LayoutRect clipRect(pixelSnappedRect);
+        if (m_imageRect) {
+            backgroundClipStateSaver.ensureSave();
+            context.clip(clipRect);
+            clipRect.inflate(1_lu);
+        }
+        geometry.clip(clipRect);
```

Comment:
BackgroundImageGeometry contains layout (unsnapped) geometry; it's wrong to clip to a paint-snapped rect here. Maybe this should clip to `rect` ?